controllers/topology/current_state.go

```diff
@@ -18,12 +18,142 @@ package topology
 
 import (
 	"context"
+	"fmt"
 
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// Gets the current state of the Cluster topology.
-func (r *ClusterReconciler) getCurrentState(ctx context.Context, cluster *clusterv1.Cluster) (*clusterTopologyState, error) {
-	// TODO: add get class logic; also remove nolint exception from clusterTopologyState and machineDeploymentTopologyState
-	return nil, nil
+// getCurrentState gets information about the current state of a Cluster by inspecting the state of the InfrastructureCluster,
+// the ControlPlane, and the MachineDeployments associated with the Cluster.
+func (r *ClusterReconciler) getCurrentState(ctx context.Context, cluster *clusterv1.Cluster, class *clusterv1.ClusterClass) (*clusterTopologyState, error) {
+	clusterState := clusterTopologyState{
+		cluster: cluster,
+	}
+
+	// Reference to the InfrastructureCluster can be nil and is expected to be on the first reconcile.
+	// In this case the method should still be allowed to continue.
+	if cluster.Spec.InfrastructureRef != nil {
+		infra, err := r.getCurrentInfrastructureClusterState(ctx, cluster)
+		if err != nil {
+			return nil, err
+		}
+		clusterState.infrastructureCluster = infra
+	}
+
+	// Reference to the ControlPlane can be nil, and is expected to be on the first reconcile. In this case the method
+	// should still be allowed to continue.
+	if cluster.Spec.ControlPlaneRef != nil {
+		cp, err := r.getCurrentControlPlaneState(ctx, cluster, class)
+		if err != nil {
+			return nil, err
+		}
+		clusterState.controlPlane = cp
+	}
+
+	// A Cluster may have zero or more MachineDeployments and a Cluster is expected to have zero MachineDeployments on
+	// first reconcile.
```

Comment:
What exactly do we mean by first reconcile? Why would the number of MachineDeployments change on nth reconcile?

Reply:
On the first reconcile the Machine Deployments won't have been spun up as this happens when the objects are initially created. The first reconcile - if things are working right - should be able to add the correct number of machine deployments in the spec and then the current state will come into line with the desired state as these objects are created and come online.